### PR TITLE
feat: add Sliced Iterative Gaussianization (SIG/GIS)

### LINKED ISSUE #68: feat: shared 1D rational-quadratic spline primitive
## Summary

Extract a shared 1D monotonic rational-quadratic (RQ) spline primitive into `rbig/_src/spline.py`. This is the core 1D transform used by both RBIG's `SplineGaussianizer` and the new SIG/GIS algorithms.

The key insight: the 1D spline operation is identical whether applied to a raw dimension (`X[:, k]`) or a projected slice (`X @ a_k`). This module provides the pure math.

### `RQSpline` class
- `fit(x_data, n_knots, kde_bw)` — estimate knots via KDE → CDF → `erfinv(2*CDF - 1)` for Gaussian target placement
- `forward(x)` → `y, log_dy_dx`
- `inverse(y)` → `x, log_dx_dy`
- Reference: sinflow's `MonotonicRationalQuadraticSpline`

### Tests (`tests/test_spline.py`)
- Forward + inverse roundtrip: `x ≈ inverse(forward(x))`
- Log-det consistency: numerical gradient vs analytical
- Monotonicity: output is strictly increasing
- Edge cases: values at/near knot boundaries

## Dependencies

- **Parent**: #23
- **Blocks**: SIGLayer (TBD), SplineGaussianizer refactor (TBD)
- **Blocked by**: nothing

## References

- [sinflow `MonotonicRationalQuadraticSpline`](https://github.com/minaskar/sinflow)
- [SINF `RQspline.py`](https://github.com/biweidai/SINF)
- [arXiv:2007.00674](https://arxiv.org/abs/2007.00674)

### PR DESCRIPTION
**Pull request by jejjohnson**
## Summary

Implements the **code half** of the Sliced Iterative Gaussianization epic (#23) — the shared primitives, the layer, the two model classes, public exports, and the `SplineGaussianizer` refactor onto the shared spline.

Closes #68, #69, #70, #71, #72, #73, #74, #81.

## What's included

| Issue | Deliverable |
|---|---|
| #68 | `rbig/_src/spline.py` — `RQSpline`: strictly monotonic rational-quadratic spline, KDE→CDF knot placement, analytic forward/inverse + log-derivative, linear tails |
| #69 | `rbig/_src/stiefel.py` — `wasserstein_1d`, `random_orthogonal_directions`, `max_sliced_wasserstein_directions` (K-SWD via a Cayley curvilinear search) |
| #70 | `rbig/_src/convergence.py` — shared `StoppingCriterion` (log-likelihood / SWD / total-correlation early stopping) |
| #71 | `SIGLayer` — projection + per-direction spline Gaussianization with a tractable log-det |
| #72 / #73 | `GIS` and `SIG` sklearn estimators sharing one greedy flow engine |
| #74 | Public exports: `GIS`, `SIG`, `SIGLayer`, `RQSpline` |
| #81 | `SplineGaussianizer` now uses the shared `RQSpline` internally; public API and existing behavior preserved |

## Testing

- New: `tests/test_spline.py`, `tests/test_stiefel.py`, `tests/test_convergence.py`, `tests/test_sliced.py` (52 tests).
- Full suite **403 passed / 1 skipped**; `ruff check .` and `ruff format --check .` both clean.
- GIS/SIG `inverse(transform(x))` roundtrip error ≈ 1e-15 (analytic spline inverse).

## Two judgment calls worth a look

1. **#70 — `AnnealedRBIG` not rewired.** The `StoppingCriterion` module supports the `total_correlation` metric so it *can* drop into `AnnealedRBIG`, but I did not rewire `AnnealedRBIG`'s inline stopping to avoid risking its existing, well-tested convergence behavior. `GIS`/`SIG` use the new module.
2. **GIS vs SIG is configuration-level.** Both are implemented as one greedy data→Gaussian flow (a symmetric bijection): `GIS` is oriented toward density estimation (`transform`/`score_samples`, `log_likelihood` stopping) and `SIG` toward generation (`sample`/`inverse_transform`, `swd` stopping). This is a simplification of the issue's literal "build layers in reverse / prepend" description, but functionally correct and sklearn-clean.

## Not included (deferred docs half of #23)

#75 (theory page) and #76–#80 (five executed notebooks) are intentionally out of scope for this PR.

https://claude.ai/code/session_01JWacugtFBatLeiZ1UW274X

---
_Generated by [Claude Code](https://claude.ai/code/session_01JWacugtFBatLeiZ1UW274X)_

### CODE REVIEW

File: rbig/_src/stiefel.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    Zs = _subsample(Z_target, n, rng)
+
+    A = random_orthogonal_directions(d, k, random_state=rng.integers(2**31))
+    obj, grad = _kswd_objective_and_grad(A, Xs, Zs)
+    history = [obj]
+    eye = np.eye(d)
+
+    for _ in range(max_iter):
+        # Skew-symmetric generator; the "+" sign ascends the objective.
+        W = grad @ A.T - A @ grad.T
```
[the remark this answers is not in the document]
Comment (jejjohnson) [reply rbig/_src/stiefel.py]: Good catch — confirmed. With `W = G·Aᵀ − A·Gᵀ` the Cayley path has `dF/dτ|₀ = −½‖W‖²`, so every step was rejected and the optimizer returned the initial random frame (verified: history length 1, objective unchanged). Fixed in 747f505 by flipping to `W = A·Gᵀ − G·Aᵀ`, which ascends. The Stiefel tests now require strict monotone ascent and a real margin over random directions, so a no-op would fail.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: rbig/_src/sliced.py [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+        if n < 2:
+            raise ValueError(f"Need at least 2 samples, got {n}.")
+        self.n_features_in_ = d
+        rng = np.random.default_rng(self.random_state)
+        k = self._resolve_directions(d)
+
+        # Optional whitening pre-transform.
+        if self.whiten:
+            self.whitener_ = PCARotation(whiten=True).fit(X)
+            Xw = self.whitener_.transform(X)
```
[the remark this answers is not in the document]
Comment (jejjohnson) [reply rbig/_src/sliced.py:262]: Fixed in 747f505: `fit` now raises a `ValueError` when `whiten=True` and `n_samples <= n_features`, with a message pointing to `whiten=False` for wide data. Added tests covering both the rejection and that wide data still fits/round-trips with whitening off.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: rbig/_src/sliced.py [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+                kde_bandwidth=self.kde_bandwidth,
+                random_state=int(rng.integers(2**31)),
+            )
+            layer.fit(R_train, X_target=target)
+            R_train, _ = layer.transform(R_train)
+            R_val, ld_val = layer.transform(R_val)
+            log_det_val += ld_val
+            self.layers_.append(layer)
+
+            if crit.update(R_val, log_det=log_det_val):
```
[the remark this answers is not in the document]
Comment (jejjohnson) [reply rbig/_src/sliced.py:302]: Fixed in 747f505: on stopping, `layers_` is truncated to `best_iter_ + 1` so the patience-window layers that worsen the selected metric are discarded. Added a test asserting `len(layers_) == best_iter_ + 1`.

---
_Generated by [Claude Code](https://claude.ai/code)_
